concept/tests/test_thing.rs

```diff
@@ -655,3 +739,203 @@ fn role_player_duplicates() {
+        attr_type
+            .set_annotation(&mut snapshot, &type_manager, AttributeTypeAnnotation::Independent(AnnotationIndependent))
+            .unwrap();
+
+        snapshot.commit().unwrap();
+        attr_type
+    };
+
+    {
+        let mut snapshot: WriteSnapshot<WALClient> = storage.clone().open_snapshot_write();
+        thing_manager
+            .create_attribute(&mut snapshot, attr_type.clone(), Value::String(Cow::Borrowed(short_string.as_str())))
+            .unwrap();
+        thing_manager
+            .create_attribute(&mut snapshot, attr_type.clone(), Value::String(Cow::Borrowed(long_string.as_str())))
+            .unwrap();
+        snapshot.commit().unwrap();
+    };
+
+    // read them back by type
+    {
+        let snapshot: WriteSnapshot<WALClient> = storage.clone().open_snapshot_write();
+        let mut attrs = thing_manager.get_attributes_in(&snapshot, attr_type.clone()).unwrap().collect_cloned();
+        let attr_values: Vec<String> = attrs
+            .into_iter()
+            .map(|mut attr| (*attr.get_value(&snapshot, &thing_manager).unwrap().unwrap_string()).to_owned())
+            .collect();
+        assert!(attr_values.contains(&short_string));
+        assert!(attr_values.contains(&long_string));
+    }
+    // read them back by value
+    {
+        let snapshot: WriteSnapshot<WALClient> = storage.clone().open_snapshot_write();
+        let mut read_short_string = thing_manager
+            .get_attribute_with_value(&snapshot, attr_type.clone(), Value::String(Cow::Borrowed(short_string.as_str())))
+            .unwrap()
+            .unwrap();
+        assert_eq!(short_string, read_short_string.get_value(&snapshot, &thing_manager).unwrap().unwrap_string());
+        let mut read_long_string = thing_manager
+            .get_attribute_with_value(&snapshot, attr_type.clone(), Value::String(Cow::Borrowed(long_string.as_str())))
+            .unwrap()
+            .unwrap();
+        assert_eq!(long_string, read_long_string.get_value(&snapshot, &thing_manager).unwrap().unwrap_string());
+    }
+}
+
+#[test]
+fn struct_write_read() {
+    let (storage, type_manager, thing_manager) = prepare();
+
+    let attr_label = Label::build("struct_test_attr");
+    let struct_name = "struct_test_test";
+    let fields: HashMap<String, (ValueType, bool)> =
+        HashMap::from([("f0l".to_owned(), (ValueType::Long, false)), ("f1s".to_owned(), (ValueType::String, false))]);
+    let definition = StructDefinition::define(struct_name.to_owned(), fields);
+
+    let instance_fields = HashMap::from([
+        ("f0l".to_owned(), FieldValue::Long(123)),
+        ("f1s".to_owned(), FieldValue::String(Cow::Owned("abc".to_owned()))),
+    ]);
+    let struct_key = {
+        let mut snapshot: WriteSnapshot<WALClient> = storage.clone().open_snapshot_write();
+        let struct_key = type_manager.create_struct(&mut snapshot, definition.clone()).unwrap();
```

Comment:
Should be it be `definition.clone()` or `&definition`? Does it need ownership? Probably serialising, so it may not?